### PR TITLE
Undo recording unification and defect fixes

### PR DESCRIPTION
**Pull request by Benglin**
Background
-----
This pull request fixes the following issues due to recent restructuring:

1. Crash during undo (this is verified by unit test `Defect_MAGN_775`):

    - Create code block node with content `a`
    - Create another with content `b`
    - Connect `a` to `b`
    - Update `b` to `b = 2` (input port is removed)
    - Undo once (**bug**: input port was not restored)
    - Undo once (**bug**: Dynamo crashes)

2. Connector disappearance after undo:

    - Create code block node with content `a`
    - Create another with content `b`
    - Connect `a` to `b`
    - Undo once (connector is removed)
    - Undo once (code block `b` is removed)
    - Redo once (code block `b` is restored)
    - Redo once (**bug**: connector is restored but invisible)

3. After fixing issue 2 above, connector disappearance after reconnection:

    - Create two `+` nodes
    - Connecting from one to another
    - Undo twice (the connector and the second `+` node are removed)
    - Redo twice (the connector and the second `+` node are restored)
    - Create another connector on top of the same connector
    - **Bug**: the connector is created but stay invisible

4. Code block node [losing focus not getting removed](https://github.com/DynamoDS/Dynamo/pull/2920):

    - Double click on canvas to create a new code block node
    - Code block awaits input with focus on its editor
    - Click on the empty space on canvas
    - **Bug**: the empty code block remains

Code Changes
-----
In addition to fixing the above issues, these are some significant code changes in the pull request:

#### Introduction of `NodeModificationUndoHelper`
This is a new `IDisposable` derived class that helps record node modification (and connector removals due to that modification) for undo. Like various *controller classes* introduced during the last major restructuring exercise, this class aims at removing undo recording from within the handlers of `ModelEventCommand` and `UpdateModelValueCommand`, which by design should not worry about undo recording. 

With this pull request, the task for undo recording is now handled at `WorkspaceModel` (which owns `UndoRedoRecorder` object) level when it translates these `RecordableCommand` objects into actual function calls on `NodeModel`. Not only it removes complexities around undo recording from the command handlers, but also unifies the way it is done to a centralized place -- `WorkspaceModel`.

#### Equality operators in `ConnectorModel`
`ConnectorModel` objects are not being identified by their `GUID` like other `ModelBase` derived classes. New operators `==` and `!=` are introduced in this pull request for `ConnectorModel` comparison.

#### Removal of empty code block nodes
[Code changes in a previous pull request](https://github.com/DynamoDS/Dynamo/pull/2920) have now been reimplemented in a different way. User can still drag an empty code block node around during its creation, and the empty code block node will be removed when user clicks on the canvas.

Unit Tests
-----
A new test case has been added to verify an area that was not covered by existing unit tests:

    - RedoDeletedNodeShowsConnector

Follow-up
-----
To avoid making this pull request too lengthy, unused `UndoRedoRecorder` parameter to these methods is not removed for now (these will be updated in a subsequent pull request):

    - ModelBase.UpdateValueCore
    - ModelBase.HandleModelEventCore

Reviewers
-----
Hi @Steell, if you do get 20% time on the other side, this may interest you (and happy birthday by the way) ;)

Hi @pboyer @ramramps, since you have originally spent quite a lot of efforts in #2920, there might be something you can say about changes specific to that area.

Thanks guys!

### CODE REVIEW

File: src/DynamoCoreWpf/Views/CodeBlocks/CodeBlockEditor.xaml.cs [hunk 1]
```diff
@@ -383,10 +384,15 @@ private void ShowInsightWindow(IEnumerable<CodeBlockInsightItem> items)
...
-            this.nodeViewModel.DynamoViewModel.ExecuteCommand(
-                   new DynCmd.UpdateModelValueCommand(
-                       this.nodeViewModel.NodeModel.GUID, "Code", this.InnerTextEditor.Text));           
+            InnerTextEditor.TextArea.ClearSelection();
+            var recorder = nodeViewModel.WorkspaceViewModel.Model.UndoRecorder;
+
+            if (string.IsNullOrEmpty(InnerTextEditor.Text))
+                DiscardChangesAndOptionallyRemoveNode(recorder);
+            else
+                CommitChanges(recorder);
```
Comment (pboyer) [inline src/DynamoCoreWpf/Views/CodeBlocks/CodeBlockEditor.xaml.cs]: Being nitpicky, but leaving out curly brackets opens up opportunities for mistakes.

File: src/DynamoCoreWpf/Views/Core/NodeView.xaml.cs [hunk 2]
```diff
@@ -319,11 +320,13 @@ private void topControl_MouseLeftButtonDown(object sender, MouseButtonEventArgs
...
-            ViewModel.DynamoViewModel.ReturnFocusToSearch();
-
-            view.mainGrid.Focus();
+            if ((ViewModel.NodeLogic is CodeBlockNodeModel) == false)
+            {
+                // Do not return focus to search if this is a code block node.
+                var view = WpfUtilities.FindUpVisualTree<DynamoView>(this);
+                ViewModel.DynamoViewModel.ReturnFocusToSearch();
+                view.mainGrid.Focus();
+            }
```
Comment (pboyer) [inline src/DynamoCoreWpf/Views/Core/NodeView.xaml.cs]: This is some yucky code, but I think this optimization makes sense.

**Comment by pboyer** [comment]
@Benglin - LGTM!

File: src/DynamoCoreWpf/Views/CodeBlocks/CodeBlockEditor.xaml.cs (hunk 1 above)
Comment (Benglin) [inline src/DynamoCoreWpf/Views/CodeBlocks/CodeBlockEditor.xaml.cs]: Thanks @pboyer, I have heard you said it before. It was my personal preference to leave them out for single-line statement, but it's about time I took this feedback seriously. Will start doing that from next time on.

File: src/DynamoCoreWpf/Views/Core/NodeView.xaml.cs (hunk 2 above)
Comment (Benglin) [inline src/DynamoCoreWpf/Views/Core/NodeView.xaml.cs]: Come to think about it, I do agree with you about this being yucky. *Keeping focus* is a UI-specific concern that should be specifiable from a `NodeView`, perhaps through some sort of property like `KeepInputFocusWhenDragged` on `INodeViewCustomization`.

**Pull request merged**